### PR TITLE
Add improved video player

### PR DESCRIPTION
**Pull request by asarium**
Uses FFmpeg for video decoding and multiple threads for reading and decoding files. Can handle 1080p with no stuttering.

This change requires the boost libraries and some prebuilt libraries. The prebuilt libraries are downloaded as part of the CMake configuration process.

There are currently no Mac builds of FFmpeg since I have no machine where I could compile those. The CI currently uses the homebrew libraries but someone will need to compile the libraries so we can include those in release builds.

### CODE REVIEW

File: cmake/platform-darwin.cmake [hunk 1]
```diff
@@ -15,3 +15,5 @@ SET(CMAKE_OSX_DEPLOYMENT_TARGET "10.9")
 SET(CMAKE_SKIP_RPATH FALSE)
 SET(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
 SET(CMAKE_INSTALL_RPATH @loader_path/../Frameworks/)
+
+set(PLATFORM_MAC TRUE CACHE INTERNAL "" FORCE)
```
Comment (chief1983) [inline cmake/platform-darwin.cmake]: Should have a newline here at the end of the file.

**Comment by SamuelCho** [comment]
Do we really need Ffmpeg to replace the video player? At the cost of adding a huge library like  boost?

**Comment by chief1983** [comment]
I had also voiced concerns earlier about its size, although not particularly because of boost.  If we're adding boost for anything, there are already several other uses for it.  It's the rest of the bulk I had wondered if there wasn't a smaller option for in this case.

**Comment by asarium** [comment]
FFmpeg is for H.264 decoding since that is superior to the Theora codec. As far as I know, the only reason I need boost is for the `sync_bounded_queue` since I don't want to implement that myself. It is very easy to make mistakes in multi threaded code (I tried, it didn't work) and boost has a working implementation. If you have an alternate solution for the bounded queue I would gladly remove the boost dependency.

**Comment by asarium** [comment]
@chief1983 FFmpeg is a proven solution that is being used in a lot of projects. Implementing our own solution for H.264 decoding via another external library will likely lead to errors since we also need to implement a parser for MP4 files.

File: code/cutscene/ffmpeg/AudioDecoder.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	explicit AudioDecoder(DecoderStatus* status);
+
+	virtual ~AudioDecoder();
+
+	virtual void decodePacket(AVPacket* packet) SCP_OVERRIDE;
+
+	virtual void finishDecoding() SCP_OVERRIDE;
+};
+}
+}
```
Comment (chief1983) [inline code/cutscene/ffmpeg/AudioDecoder.h]: Need a newline here too

File: code/cutscene/ffmpeg/VideoDecoder.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	explicit VideoDecoder(DecoderStatus* status);
+
+	virtual ~VideoDecoder();
+
+	virtual void decodePacket(AVPacket* packet) SCP_OVERRIDE;
+
+	virtual void finishDecoding() SCP_OVERRIDE;
+};
+}
+}
```
Comment (chief1983) [inline code/cutscene/ffmpeg/VideoDecoder.h]: And a newline here.

File: code/cutscene/ffmpeg/internal.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+	if (audioCodecCtx != nullptr) {
+		avcodec_close(audioCodecCtx);
+#if LIBAVCODEC_VERSION_INT > AV_VERSION_INT(57, 24, 255)
+		avcodec_free_context(&audioCodecCtx);
+#endif
+		audioCodecCtx = nullptr;
+	}
+}
+}
```
Comment (chief1983) [inline code/cutscene/ffmpeg/internal.cpp]: When braces stack like this, could we use comments indicating what is closed where to increase clarity?

**Comment by SamuelCho** [comment]
Does ffmpeg require multiple threads to be running? Or is extracting good performance out of video decoding contingent on having multithreading?

**Comment by asarium** [comment]
It's for getting better performance. The current ogg player is single threaded and that is barely able to play a 720p cutscene.

**Comment by SamuelCho** [comment]
I see, but why a synchronized bounded queue? Would a lock free queue not work?

**Comment by asarium** [comment]
I need the ability to limit how many frames are waiting in the queue. That is required since otherwise we would just be wasting memory and CPU time with decoding frames we may not even need.

**Comment by SamuelCho** [comment]
Is there any particular reason why we can't use a smaller library like this one? It's meant for single-producer, single-consumer problems which seems to be the case here. Apparently the author used it for audio processing code, very similar to what this is trying to do. And the queue can be bounded or dynamic. https://github.com/cameron314/readerwriterqueue

**Comment by asarium** [comment]
The library looks promising. I'll do some tests to see if it will work with my code and if it does then I should be able to remove the boost dependency. The only reason I added boost in the first place is because I needed a threading library and `std::thread` wasn't available at that point.

**Comment by asarium** [comment]
@SamuelCho The library you linked to doesn't work for this since it doesn't allow to specify an upper limit for the number of elements. It just allows to specify an initial capacity but if more elements are added the queue grows. I'll continue searching for other alternatives but at the moment it looks like boost is the best option.

**Comment by SamuelCho** [comment]
Try_enqueue didn't work?

**Comment by asarium** [comment]
`try_enqueue` immediately returns if the capacity is reached but I need a function that waits until there is room in the queue. The blocking queue that exists only supports that for the removing elements from the queue.

**Comment by SamuelCho** [comment]
What's stopping you from just putting a condition variable that waits if the queue is full? Sorry if this sounds like backseat driving.

**Comment by asarium** [comment]
Then I would just be reimplementing what the boost queue already does and I don't like duplicating work, especially if it's error-prone like multi threading stuff. If there is a solution that can be used without major changes to the code then I'm willing to make the changes but ignoring an existing solution only to avoid including boost is not very efficient.

**Comment by asarium** [comment]
It's not a minor change since I need to make sure that the variable is signaled when the queue isn't full anymore by the function that removes the frames but it also needs to make sure that there are no edge cases where the synchronization or notification may fail. This all requires effort that I am not willing to invest if there is a perfectly fine solution already available.

**Comment by The-E** [comment]
Way I see it, we want this feature. The best option to implement it, according to the expert on the subject, involves using a boost feature. So, we either use boost, or we do not use this feature.

I say we use this. This discussion has gotten borderline ridiculous, and I have seen no good reasons against the use of boost in it.

**Comment by SamuelCho** [comment]
Okay, sorry for pushing back so hard. We kind of have boost allergy in retail game dev land. It's a hard habit to break.

**Comment by asarium** [comment]
@chief1983 and I are currently working on getting Mac libraries for FFmpeg compiled and integrated into our system. After that is done this should be ready to be merged.

**Comment by asarium** [comment]
The libraries @chief1983 has provided seem to be working fine. They are now used by Travis CI but I can't confirm if the generated executable actually runs.

**Comment by asarium** [comment]
From my perspective, these changes are finished and are ready to be merged. Is there anything else that should be changed or discussed?

**Comment by chief1983** [comment]
Got the trimmed-down ffmpeg lib for Mac uploaded, and confirmed that the boost static linking works fine.  If we can get that lib updated, I don't see any issue with this from a mac standpoint.  Ogg videos are still playing.  Anyone have the MVE files to test on Mac?

**Comment by asarium** [comment]
I'm currently recompiling the Linux and Windows libraries with the updated flags. MVE is uses the old code so there shouldn't be any changes. I'll see if I have the MVE cutscenes to test if they still work.

**Comment by asarium** [comment]
So, it turns out that FFmpeg is actually able to decode the old MVE videos so I replaced the old code with the new FFmpeg decoder. This should also fix #880 since the new display code works regardless of which decoder is used.

**Comment by chief1983** [comment]
This makes a perfectly usable Mac build now, with filesizes not terribly exceeding what we've tolerated in the past when we weren't copying in SDL correctly.  Since no one objected then, no one's likely to now, and I don't think can trim down the new boost/ffmpeg dependencies much more unless we want to pick out individual supported formats from ffmpeg.  So, if someone wants to pass this for general code review, it's got my sign-off.

**Pull request merged**